exonum-java-binding-common/src/test/java/com/exonum/binding/common/proofs/list/ListProofHashCodeCalculatorTest.java

```diff
@@ -0,0 +1,176 @@
+/*
+ * Copyright 2018 The Exonum Team
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.exonum.binding.common.proofs.list;
+
+import static com.exonum.binding.common.hash.Funnels.hashCodeFunnel;
+import static com.google.common.collect.ImmutableMap.of;
+import static org.hamcrest.CoreMatchers.equalTo;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNotEquals;
+
+import com.exonum.binding.common.hash.HashCode;
+import com.exonum.binding.common.hash.Hashing;
+import com.exonum.binding.common.hash.PrimitiveSink;
+import com.exonum.binding.common.serialization.StandardSerializers;
+import java.nio.charset.Charset;
+import org.junit.jupiter.api.Test;
+
+class ListProofHashCodeCalculatorTest {
+
+  private static final String V1 = "v1";
+  private static final String V2 = "v2";
+  private static final String V3 = "v3";
+  private static final String V4 = "v4";
+
+  private static final HashCode H1 = HashCode.fromString("a1");
+  private static final HashCode H2 = HashCode.fromString("a2");
+
+  private ListProofRootHashCalculator<String> calculator;
+
+  @Test
+  void visit_SingletonListProof() {
+    ListProof root = leafOf(V1);
+
+    calculator = createListProofCalculator();
+    root.accept(calculator);
+
+    assertThat(calculator.getElements(), equalTo(of(0L, V1)));
+    assertEquals(getNodeHashCode(V1), calculator.getCalculatedRootHash());
+  }
+
+  @Test
+  void visit_FullProof2elements() {
+    ListProofElement left = leafOf(V1);
+    ListProofElement right = leafOf(V2);
+    ListProofBranch root = new ListProofBranch(left, right);
+
+    calculator = createListProofCalculator();
+    calculator.visit(root);
+
+    //calculate expected root hash
+    HashCode expectedRootHash = getBranchHashCode(getNodeHashCode(V1), getNodeHashCode(V2));
+
+    assertThat(calculator.getElements(), equalTo(of(0L, V1,
+        1L, V2)));
+    assertEquals(expectedRootHash, calculator.getCalculatedRootHash());
+  }
+
+  @Test
+  void visit_FullProof4elements() {
```

Comment:
I'd add a test for a 3-element tree (that has a branch node with a _single_ child).